File: src/hover.ts
// Hover / quick-info rendering for a resolved symbol. Shared by the LSP server
// and the fourslash hover baselines so both render identically.

import type { Checker } from "./checker.ts";
import { type Symbol, SymbolFlags } from "./types.ts";

export function symbolKindWord(flags: SymbolFlags): string {
  if (flags & SymbolFlags.Class) return "class";
  if (flags & SymbolFlags.Interface) return "interface";
  if (flags & SymbolFlags.Enum) return "enum";
  if (flags & SymbolFlags.Record) return "record";
  if (flags & SymbolFlags.Annotation) return "@interface";
  if (flags & SymbolFlags.Constructor) return "constructor";
  if (flags & SymbolFlags.Method) return "method";
  if (flags & SymbolFlags.Field) return "field";
  if (flags & SymbolFlags.EnumConstant) return "enum constant";
  if (flags & SymbolFlags.Parameter) return "parameter";
  if (flags & SymbolFlags.TypeParameter) return "type parameter";
  if (flags & SymbolFlags.LocalVariable) return "local variable";
  return "symbol";
}

const TYPE_FLAGS =
  SymbolFlags.Class |
  SymbolFlags.Interface |
  SymbolFlags.Enum |
  SymbolFlags.Record |
  SymbolFlags.Annotation;

/**
 * One-line hover label, in the style of the C#/Roslyn language service:
 *   methods  -> the full signature            e.g.  int add(int a, int b)
 *   types    -> keyword + name                e.g.  class Foo
 *   values   -> (kind) type name              e.g.  (field) int count
 *   type var -> (type parameter) name         e.g.  (type parameter) T
 */
export function getHoverText(checker: Checker, symbol: Symbol): string {
  if (symbol.flags & (SymbolFlags.Method | SymbolFlags.Constructor)) {
    const signature = checker.signatureOfSymbol(symbol);
    if (signature) return signature;
  }
  const word = symbolKindWord(symbol.flags);
  if (symbol.flags & TYPE_FLAGS) {
    return `${word} ${symbol.escapedName}`;
  }
  if (symbol.flags & SymbolFlags.TypeParameter) {
    return `(${word}) ${symbol.escapedName}`;
  }
  const type = checker.typeStringOfSymbol(symbol);
  // Omit an unresolvable type (e.g. a lambda parameter whose target is unknown)
  // rather than printing "<error>".
  return type === "<error>"
    ? `(${word}) ${symbol.escapedName}`
    : `(${word}) ${type} ${symbol.escapedName}`;
}

File: togo/internal/services/hover.go
package services

// Hover / quick-info rendering for a resolved symbol. Shared by the LSP server
// and the fourslash hover baselines so both render identically.
// Port of src/services/hover.ts.

import (
	"strings"

	"github.com/nikeee/cappu/internal/compiler"
)

// SymbolKindWord returns the human-readable kind word for a symbol's flags.
func SymbolKindWord(flags compiler.SymbolFlags) string {
	switch {
	case flags&compiler.SymbolFlagsPackage != 0:
		return "package"
	case flags&compiler.SymbolFlagsClass != 0:
		return "class"
	case flags&compiler.SymbolFlagsInterface != 0:
		return "interface"
	case flags&compiler.SymbolFlagsEnum != 0:
		return "enum"
	case flags&compiler.SymbolFlagsRecord != 0:
		return "record"
	case flags&compiler.SymbolFlagsAnnotation != 0:
		return "@interface"
	case flags&compiler.SymbolFlagsConstructor != 0:
		return "constructor"
	case flags&compiler.SymbolFlagsMethod != 0:
		return "method"
	case flags&compiler.SymbolFlagsField != 0:
		return "field"
	case flags&compiler.SymbolFlagsEnumConstant != 0:
		return "enum constant"
	case flags&compiler.SymbolFlagsParameter != 0:
		return "parameter"
	case flags&compiler.SymbolFlagsTypeParameter != 0:
		return "type parameter"
	case flags&compiler.SymbolFlagsLocalVariable != 0:
		return "local variable"
	default:
		return "symbol"
	}
}

const typeFlags = compiler.SymbolFlagsClass | compiler.SymbolFlagsInterface |
	compiler.SymbolFlagsEnum | compiler.SymbolFlagsRecord | compiler.SymbolFlagsAnnotation

// EnclosingCall returns the call whose callee is this identifier (directly or
// via recv.name), or nil.
func EnclosingCall(identifier *compiler.Node) *compiler.Node {
	parent := identifier.Parent
	if parent.Kind == compiler.CallExpression && parent.AsCallExpression().Expression == identifier {
		return parent
	}
	if parent.Kind == compiler.PropertyAccessExpression && parent.AsPropertyAccessExpression().Name == identifier &&
		parent.Parent.Kind == compiler.CallExpression && parent.Parent.AsCallExpression().Expression == parent {
		return parent.Parent
	}
	return nil
}

func typeParameterBounds(symbol *compiler.Symbol) string {
	declaration := symbolDeclaration(symbol)
	if declaration == nil || declaration.Kind != compiler.TypeParameter {
		return ""
	}
	bounds := declaration.AsTypeParameter().Constraint
	if bounds == nil || bounds.Len() == 0 {
		return ""
	}
	text := compiler.GetSourceFileOfNode(declaration).AsSourceFile().Text
	var parts []string
	for _, b := range bounds.Nodes {
		parts = append(parts, text[compiler.SkipTrivia(text, b.Pos):b.End])
	}
	return strings.Join(parts, " & ")
}

// GetHoverText renders a one-line hover label for a resolved symbol. atNode is
// the referencing identifier when hovering a use rather than the declaration
// (pass nil for the declaration).
func GetHoverText(checker *compiler.Checker, symbol *compiler.Symbol, atNode *compiler.Node) string {
	if symbol.Flags&(compiler.SymbolFlagsMethod|compiler.SymbolFlagsConstructor) != 0 {
		var call *compiler.Node
		if atNode != nil && atNode.Kind == compiler.Identifier {
			call = EnclosingCall(atNode)
		}
		if call != nil {
			if instantiated, ok := checker.InstantiatedSignatureOfCall(call); ok {
				return instantiated
			}
		}
		if signature, ok := checker.SignatureOfSymbol(symbol); ok {
			return signature
		}
	}
	word := SymbolKindWord(symbol.Flags)
	if symbol.Flags&(typeFlags|compiler.SymbolFlagsPackage) != 0 {
		return word + " " + symbol.EscapedName
	}
	if symbol.Flags&compiler.SymbolFlagsTypeParameter != 0 {
		bounds := typeParameterBounds(symbol)
		if bounds != "" {
			return "(" + word + ") " + symbol.EscapedName + " extends " + bounds
		}
		return "(" + word + ") " + symbol.EscapedName
	}
	typ := useSiteTypeString(checker, atNode)
	if typ == "" {
		typ = checker.TypeStringOfSymbol(symbol)
	}
	if typ == "<error>" {
		return "(" + word + ") " + symbol.EscapedName
	}
	return "(" + word + ") " + typ + " " + symbol.EscapedName
}

func useSiteTypeString(checker *compiler.Checker, atNode *compiler.Node) string {
	if atNode == nil || atNode.Kind != compiler.Identifier ||
		atNode.Parent == nil || atNode.Parent.Kind != compiler.PropertyAccessExpression ||
		atNode.Parent.AsPropertyAccessExpression().Name != atNode {
		return ""
	}
	t := checker.GetTypeOfExpression(atNode.Parent)
	if t.Kind == compiler.TypeKindError {
		return ""
	}
	return compiler.TypeToString(t)
}
